File: tests/benchmarks/coro/fanout/fanout.erl
%% Fanout benchmark - Erlang equivalent
%% Dispatcher sends tasks to N workers via shared broker, collector gathers results
%% Usage: erl -noshell -s fanout main -s init stop
-module(fanout).
-export([main/0, main/2]).

main() ->
    main(1000, 100000).

main(N, Tasks) ->
    io:format("=== Fanout 测试 (Erlang) ===~n"),
    io:format("工作协程: ~p~n", [N]),
    io:format("任务数量: ~p~n", [Tasks]),

    Self = self(),
    T0 = erlang:monotonic_time(millisecond),

    %% Task channel (simulated by broker)
    TaskCh = spawn(fun() -> task_broker(queue:new(), [], 0) end),

    %% Start N workers, each loops: recv from TaskCh, send result to Self
    [spawn(fun() -> worker(TaskCh, Self) end) || _ <- lists:seq(1, N)],

    %% Distribute tasks
    [TaskCh ! {send, I} || I <- lists:seq(0, Tasks - 1)],

    %% Signal close after all tasks sent
    TaskCh ! close,

    %% Collect results
    Sum = collect_results(Tasks, 0),

    T1 = erlang:monotonic_time(millisecond),
    Elapsed = T1 - T0,
    Expected = Tasks * (Tasks - 1),
    io:format("结果: ~p~n", [Sum]),
    io:format("预期: ~p~n", [Expected]),
    io:format("正确: ~p~n", [Sum =:= Expected]),
    io:format("时间: ~p ms~n", [Elapsed]),
    case Elapsed > 0 of
        true ->
            io:format("吞吐量: ~p tasks/sec~n", [Tasks * 1000 div Elapsed]);
        false ->
            io:format("吞吐量: 太快无法测量~n")
    end.

worker(TaskCh, Collector) ->
    TaskCh ! {recv, self()},
    receive
        {item, Val} ->
            Collector ! Val * 2,
            worker(TaskCh, Collector);
        closed ->
            ok
    end.

task_broker(Queue, Waiters, Closed) ->
    receive
        {send, Item} ->
            case Waiters of
                [W | Rest] ->
                    W ! {item, Item},
                    task_broker(Queue, Rest, Closed);
                [] ->
                    task_broker(queue:in(Item, Queue), [], Closed)
            end;
        {recv, Consumer} ->
            case queue:out(Queue) of
                {{value, Item}, Q2} ->
                    Consumer ! {item, Item},
                    task_broker(Q2, Waiters, Closed);
                {empty, _} when Closed =:= 1 ->
                    Consumer ! closed,
                    task_broker(Queue, Waiters, Closed);
                {empty, _} ->
                    task_broker(Queue, Waiters ++ [Consumer], Closed)
            end;
        close ->
            [W ! closed || W <- Waiters],
            task_broker(Queue, [], 1)
    end.

collect_results(0, Sum) ->
    Sum;
collect_results(N, Sum) ->
    receive Val ->
        collect_results(N - 1, Sum + Val)
    end.

File: tests/benchmarks/coro/fanout/fanout.go
// 作者：xingleixu@gmail.com
// 扇出扇入测试（Fan-out/Fan-in）
// 一个分发器将任务分发给 N 个工作 goroutine，然后收集结果

package main

import (
	"fmt"
	"os"
	"strconv"
	"sync"
	"time"
)

func main() {
	N := 1000       // 工作协程数量
	TASKS := 100000 // 任务数量

	if len(os.Args) > 1 {
		if n, err := strconv.Atoi(os.Args[1]); err == nil {
			N = n
		}
	}
	if len(os.Args) > 2 {
		if t, err := strconv.Atoi(os.Args[2]); err == nil {
			TASKS = t
		}
	}

	fmt.Println("=== Fanout 测试 ===")
	fmt.Println("工作协程:", N)
	fmt.Println("任务数量:", TASKS)

	taskCh := make(chan int, 100)
	resultCh := make(chan int, 100)

	var wg sync.WaitGroup

	start := time.Now()

	// 启动工作协程
	for i := 0; i < N; i++ {
		wg.Add(1)
		go func(id int) {
			defer wg.Done()
			for task := range taskCh {
				// 模拟计算
				result := task * 2
				resultCh <- result
			}
		}(i)
	}

	// 启动收集器
	var sum int64
	done := make(chan bool)
	go func() {
		for i := 0; i < TASKS; i++ {
			sum += int64(<-resultCh)
		}
		done <- true
	}()

	// 分发任务
	for i := 0; i < TASKS; i++ {
		taskCh <- i
	}

	// 关闭任务队列
	close(taskCh)

	// 等待工作协程完成
	wg.Wait()

	// 等待收集器完成
	<-done

	elapsed := time.Since(start)
	elapsedMs := float64(elapsed) / float64(time.Millisecond)

	// 预期结果: (0 + 1 + 2 + ... + (TASKS-1)) * 2 = TASKS * (TASKS - 1)
	expected := int64(TASKS) * int64(TASKS-1)

	fmt.Println("结果:", sum)
	fmt.Println("预期:", expected)
	fmt.Println("正确:", sum == expected)
	fmt.Printf("时间: %.3f ms\n", elapsedMs)
	if elapsed > 0 {
		fmt.Println("吞吐量:", int(float64(TASKS)/elapsed.Seconds()), "tasks/sec")
	} else {
		fmt.Println("吞吐量: 太快无法测量")
	}
}
